File: PersisteniadeObjetos/Ej3/main.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

class Cliente implements Serializable {
    private int id;
    private String nombre;
    private int telefono;

    public Cliente(int id, String nombre, int telefono) {
        this.id = id;
        this.nombre = nombre;
        this.telefono = telefono;
    }

    // Getters
    public int getId() { return id; }
    public String getNombre() { return nombre; }
    public int getTelefono() { return telefono; }

    @Override
    public String toString() {
        return "Cliente [ID: " + id + ", Nombre: " + nombre + ", Teléfono: " + telefono + "]";
    }
}

class ArchivoCliente {
    private String nomA;

    public ArchivoCliente(String nomA) {
        this.nomA = nomA;
    }

    public void crearArchivo() {
        try {
            new File(nomA).createNewFile();
        } catch (IOException e) {
            System.out.println("Error al crear el archivo: " + e.getMessage());
        }
    }

    public void guardarCliente(Cliente cliente) {
        List<Cliente> clientes = leerClientes();
        clientes.add(cliente);
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(nomA))) {
            oos.writeObject(clientes);
        } catch (IOException ex) {
            System.out.println("Error al guardar cliente: " + ex.getMessage());
        }
    }

    private List<Cliente> leerClientes() {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nomA))) {
            return (List<Cliente>) ois.readObject();
        } catch (EOFException e) {
            return new ArrayList<>(); // Archivo vacío
        } catch (IOException | ClassNotFoundException ex) {
            System.out.println("Error al leer clientes: " + ex.getMessage());
            return new ArrayList<>();
        }
    }

    public Cliente buscarCliente(int id) {
        List<Cliente> clientes = leerClientes();
        for (Cliente c : clientes) {
            if (c.getId() == id) {
                return c;
            }
        }
        return null;
    }

    public Cliente buscarCelularCliente(int telefono) {
        List<Cliente> clientes = leerClientes();
        for (Cliente c : clientes) {
            if (c.getTelefono() == telefono) {
                return c;
            }
        }
        return null;
    }

    public static void main(String[] args) {
        ArchivoCliente archivo = new ArchivoCliente("clientes.dat");
        archivo.crearArchivo();

        // Guardar clientes
        archivo.guardarCliente(new Cliente(1, " Belen Alvarez", 63456789));
        archivo.guardarCliente(new Cliente(2, "Luis Gómez", 72555678));
        archivo.guardarCliente(new Cliente(3, "María López", 62512345));
        archivo.guardarCliente(new Cliente(4, "Carlos Ruiz", 71234567));
        archivo.guardarCliente(new Cliente(5, "Laura Torres", 63098765));
        archivo.guardarCliente(new Cliente(6, "Pedro Martínez", 67070123));
        archivo.guardarCliente(new Cliente(7, "Sofía Jiménez", 62585756));


        // Buscar por ID
        Cliente buscado = archivo.buscarCliente(5);
        System.out.println("Cliente encontrado por ID: " + buscado);

        // Buscar por teléfono
        Cliente porCelular = archivo.buscarCelularCliente(62585756);
        System.out.println("Cliente encontrado por teléfono: " + porCelular);
    }
}

File: PersisteniadeObjetos/Ej3/main.py
import csv

class Cliente:
    def __init__(self, id: int, nombre: str, telefono: int):
        self.id = id
        self.nombre = nombre
        self.telefono = telefono

    def __str__(self):
        return f"Cliente [ID: {self.id}, Nombre: {self.nombre}, Teléfono: {self.telefono}]"

class ArchivoCliente:
    def __init__(self, nom_archivo: str):
        self.nom_archivo = nom_archivo

    def crear_archivo(self):
        try:
            with open(self.nom_archivo, 'x') as file:
                pass  # Crea el archivo vacío
        except FileExistsError:
            pass

    def guardar_cliente(self, cliente: Cliente):
        with open(self.nom_archivo, 'a', newline='') as file:
            writer = csv.writer(file)
            writer.writerow([cliente.id, cliente.nombre, cliente.telefono])

    def leer_clientes(self) -> list[Cliente]:
        clientes = []
        try:
            with open(self.nom_archivo, 'r') as file:
                reader = csv.reader(file)
                for row in reader:
                    if row:  # Evita líneas vacías
                        clientes.append(Cliente(int(row[0]), row[1], int(row[2])))
        except FileNotFoundError:
            pass
        return clientes

    def buscar_cliente(self, id: int) -> Cliente | None:
        clientes = self.leer_clientes()
        for cliente in clientes:
            if cliente.id == id:
                return cliente
        return None

    def buscar_celular_cliente(self, telefono: int) -> Cliente | None:
        clientes = self.leer_clientes()
        for cliente in clientes:
            if cliente.telefono == telefono:
                return cliente
        return None


archivo = ArchivoCliente("clientes.txt")
archivo.crear_archivo()

# Guardar clientes
archivo.guardar_cliente(Cliente(1, "Ana Pérez", 63040711))
archivo.guardar_cliente(Cliente(2, "Luis Gómez", 71234567))
archivo.guardar_cliente(Cliente(3, "María López", 5555678))
archivo.guardar_cliente(Cliente(4, "Carlos Ruiz", 98765432))
archivo.guardar_cliente(Cliente(5, "Laura Torres", 12345678))

# Buscar por ID
print("Cliente encontrado:", archivo.buscar_cliente(1))

# Buscar por teléfono
print("Cliente por teléfono:", archivo.buscar_celular_cliente(63040711))
